core/src/main/kotlin/org/jordillonch/kes/cqrs/query/domain/QueryBus.kt

```diff
@@ -1,15 +1,15 @@
 package org.jordillonch.kes.cqrs.query.domain
 
-interface Query
+interface Query<R>
 
-interface QueryHandler<in Q : Query, R> {
+interface QueryHandler<R, Q: Query<R>> {
     fun on(query: Q): R
 }
 
 interface QueryBus {
-    fun <Q : Query> registerHandler(handler: QueryHandler<Q, *>)
-    fun <Q : Query, R> registerHandler(handler: (Q) -> R)
-    fun <R> ask(query: Query): R
+    fun <R, Q: Query<R>> registerHandler(handler: QueryHandler<R, Q>)
+    fun <R, Q: Query<R>> registerHandler(handler: (Q) -> R)
```

Comment:
Thank you for taking a look to my PR 😄 

About of having the registration handler feature in the contract of the bus I feel it is part of the lifecycle of a bus so you I want to have it for clarity sake.

Also I need to use it in my saga implementation (not for query bus, for command and event bus)

Reply:
👌

I'll take a look a what you have about sagas, I have very limited experience with them and I'm curious about your take on them.

BTW, I like being an asshole... 😄 so here I go: I think using `Simple` is an easy escape and no different from using an `Impl` suffix. _Change_my_mind.jpeg_

How would a `Complex` one look like? Is `Simple` indicate of a property of the implementation or is it because the interface is already taken and we need something else? (I personally made that same question myself and that's why I ended up with `KodeinBus`.)

Reply:
Also wanted to ask about the segregation between QueryBus and CommandBus. What's your opinion about merging the functionality into a single `Bus`?

Reply:
😂 asshole would be translated as "primmirat" in this case. I think you are right. Maybe it should be called `Sync` or something like that. 
BTW for me it's not a problem to catch constructive comments. So be as ~asshole~ direct as you want 😂😂😂 

About the segregation of the bus. I think it is slightly better to segregate because:
- When you are injecting the bus in a service you can see clearly what is your intention. For instance if you inject the command bus in a domain service for me it's a bad smell. I know that the important point is if you are sending commands or queries and not the bus by himself. That's the reason because I said "slightly".
- I have the feeling that maybe I would need a different implementation for command bus and query bus but since now I always used almost the same implementation. I know that some people added more features to the query bus, for example, to be able to subscribe a changes on a query.
- Finally, I like to split it and work with them. I believe it helps to maintain the project more organized.

Reply:
Thanks for the info!